### DISCUSSION TITLE
migrate from php Yii 2 to Nestjs/Mikro orm

### DISCUSSION THREAD

**Discussion by Ahmadjlallad** [mikro-orm/mikro-orm#4984] [Q&A] [answered]
Hello, everyone!

We are currently working on a telehealth SaaS application developed with PHP Yii2 (my favorite framework). Our goal is to migrate the doctor part to NestJS. We have set up the environment and generated entities using the Entity Generator. The generated code looks like this:
for example
```ts
@Entity()
export class Consultation {

    @ManyToOne({
        entity: () => Member,
        fieldName: 'member_id',
        index: 'member_id',
        serializedName: 'member_id',
    })
    memberId!: Member;
}
```
In this example, the ```memberId``` is the ID, and we're wondering if this is the intended way or if we should structure it like this:

```ts
export class Consultation {

    @Property({ serializedName: 'member_id' })
    memberId!: number;

    @ManyToOne({
        entity: () => Member,
        serializedName: 'member',
    })
    member!: Member;
}
```
Is it correct to have the property and the relationship structured this way, or should we follow a different approach?

Thank you for your insights!

**Comment by B4nan** [comment]
> The generated code looks like this:

The ORM generator won't produce `serializedName` option, and it [should generate this instead](https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/entity-generator/__snapshots__/EntityGenerator.test.ts.snap#L89-L90):

```ts
@Entity()
export class Consultation {

    @ManyToOne({
        entity: () => Member,
        index: 'member_id', // do you have indexes named the same as the columns? looks weird but I guess you do if it generates this
    })
    member!: Member; // not memberId
}
```

That is what you will find all over the docs, as that is the suggested way of doing things. Note that you can access the FK without loading the target entity, as well as check if the relation is loaded or not. You can also opt-in to generate such relations with [`Ref` wrapper](https://mikro-orm.io/docs/next/guide/type-safety#reference-wrapper) for added type-safety (the option is called `identifiedReferences` but I guess I should rename it in v6 as this is nowadays called `Ref` wrapper).

> Is it correct to have the property and the relationship structured this way,

This is not fully supported, nor needed, you should be fine with just the relation property. But if you really want to have the extra scalar property, you can mark it as `persist: false`.

```ts
export class Consultation {

    @Property({ serializedName: 'member_id', persist: false })
    memberId!: number;

    @ManyToOne({
        entity: () => Member,
//        serializedName: 'member', // no need for this
    })
    member!: Member;
}
```

Note that this will help only for reading purposes, such property is never considered for insert/update queries.

**Reply by Ahmadjlallad** [reply]
Thank you so much

> The ORM generator won't produce serializedName option, and it [should generate this instead](https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/entity-generator/__snapshots__/EntityGenerator.test.ts.snap#L89-L90):

I modified the ORM generator script so that the serializedName will be in underscore case. This is because we have a mobile app, and we don't want to maintain two projects until all users have installed/updated to the latest version. (I know we could achieve this without adding serializedName, but it's the easiest way for now. We are still testing/experimenting with Nestjs/Mikro ORM, and you are welcome to suggest a solution with underscore to camel case and vice versa).

I want to ask if I could mark persist: false for the relation, not the scalar. For example, we have a table to represent the doctor called Location (old table used to represent the clinic location but now it represent doctor and clinic location, and we are in the restructure/normalization process). In the Location table, we have member_id as a foreign key (FK). I want to define something like this:

```ts
@Entity()
export class Member {
    @PrimaryKey()
    memberId!: number
    // ... 
}

@Entity()
export class Location {
    @PrimaryKey({
        serializedName: 'location_id',
        fieldName: 'location_id',
    })
    locationId!: number

    @Property({
        serializedName: 'member_id',
        persist: false,
    })
    memberId?: number

    @ManyToOne({
        entity: () => Member,
        fieldName: 'member_id',
        index: 'location_member_id_fk',
    })
    member?: Member
}
```

I want to add a relation to the ```Location``` from the ```Member``` (I know it does not make any sense :), but it's for easy access and join statements):

```ts


@Entity()
export class Member {
    @PrimaryKey()
    memberId!: number
    // ... 

    @OneToOne({
        entity: () => Location,
        nullable: true,
        joinColumn: 'member_id',
        referenceColumnName: 'member_id',
        persist: false,
    })
    location?: Location
}
```

It does work, but the select statement looks like this:
```sql
select `member_id`,
       `m0`.`phone`,
       `m0`.`language`,
       `l1`.`location_id`                  as `l1__location_id`,
       `l1`.`member_id`                    as `l1__member_id`
from `member` as `m0`
         left join `location` as `l1` on `m0`.`member_id` = `l1`.`member_id`
where `m0`.member_id = 123;
```
The member_id is ambiguous. Do you have any idea how to overcome this issue, or should we just query the location separately?

**Reply by B4nan** [reply, marked as answer]
> I want to ask if I could mark persist: false for the relation, not the scalar.

Not really, this would break schema diffing completely, as you wouldn't have any FKs - the scalar does not represent a FK constraint. If you don't care about using the schema generator (or migrations - it's used for generating the diffs), it should be fine.

> It does work, but the select statement looks like this:

Yes, another reason why it's a bad idea. Virtual properties (those marked with `persist: false`) won't have an alias - because they are not supposed to be represent any real database column in the first place.

I'd like to deal with this at some point, as using `persist: false` is (and always was) a workaround for an unsupported feature. 

I'd still suggest you keep only the relation property, that is the suggested way.

> I modified the ORM generator script so that the serializedName will be in underscore case.

So you'd like to work with camelCase properties but serialize them to snake_case? I guess we could add a new method to the naming strategy so you can control this globally.

**Reply by Ahmadjlallad** [reply]
Thank you for the great answer. I was exploring Event hooks and subscriptions. Is there a way to insert or update without triggering an event? Can I use native update/insert and still have the unit of work updated?

**Reply by B4nan** [reply]
I guess you could use `em.insert` or `em.nativeUpdate` (or query builder), and afterwards call `em.merge()` which updates the UoW.

**Reply by Ahmadjlallad** [reply]
Last question: I've successfully connected to both MySQL and MongoDB. However, when I attempt to serialize a MySQL entity, I encounter the following error. It seems that the Mikro ORM serializer is attempting to use MongoDB date formatting:


```txt
[Nest] 68836 - 12/13/2023, 12:15:31 PM ERROR [ExceptionsHandler] Error: RangeError: Invalid time value
    at Date.toISOString (<anonymous>)
    at Date.MongoConnection.Date.<computed> (/Users/ahmadjallad/Sites/service-doctors/node_modules/@mikro-orm/mongodb/MongoConnection.js:16:37)
    at formatValue (node:internal/util/inspect:805:19)
    at formatProperty (node:internal/util/inspect:1926:11)
    at formatRaw (node:internal/util/inspect:1054:9)
    at formatValue (node:internal/util/inspect:840:10)
    at inspect (node:internal/util/inspect:364:10)
    at Member.<computed>._a.<computed> (/Users/ahmadjallad/Sites/service-doctors/node_modules/@mikro-orm/core/entity/EntityHelper.js:110:44)
    at formatValue (node:internal/util/inspect:805:19)
    at formatProperty (node:internal/util/inspect:1926:11)
This is caused by either a bug in Node.js or incorrect usage of Node.js internals. Please open an issue with this stack trace at https://github.com/nodejs/node/issues
```

here is the configuration

```ts
@Module({
    imports: [
        MikroOrmModule.forRoot({
            dbName: process.env.MONGO_DB_NAME,
            clientUrl: process.env.MONGO_CONNECTION,
            entities: ['./dist/src/mongo/entities/'],
            entitiesTs: ['./src/mongo/entities'],
            driver: MongoDriver,
            contextName: 'mongodb',
            debug: true,
            forceUtcTimezone: true,
            registerRequestContext: false,
        }),
        MikroOrmModule.forRoot({
            dbName: process.env.DATABASE_NAME,
            driver: MySqlDriver,
            user: process.env.DATABASE_USERNAME,
            password: process.env.DATABASE_PASSWORD,
            host: process.env.DATABASE_HOST,
            entities: ['./dist/src/entities/'],
            entitiesTs: ['./src/entities'],
            debug: true,
            allowGlobalContext: true,
            namingStrategy: UnderscoreNamingStrategy,
            cache: { enabled: true },
            metadataProvider: TsMorphMetadataProvider,
            discovery: {
                alwaysAnalyseProperties: false, // do not analyse properties when not needed (with ts-morph)
            },
            loadStrategy: LoadStrategy.JOINED,
            resultCache: {
                // following options are the defaults
                adapter: MemoryCacheAdapter,
                expiration: 1000, // 1s
                options: {},
                // cache can be also enabled globally
                global: 50, // 50ms
            },
            forceUtcTimezone: true,
        }),
    ],
    controllers: [AppController],
    providers: [AppService],
})
export class AppModule implements NestModule {
}

```

Note: I didn't add ```contextName``` to MySQL because I wanted it to be injected without a token.

**Reply by B4nan** [reply]
I'd have to see a complete repro for this, but the problem here is not serialization, it's logging, you must be doing `console.log` somewhere and that is what triggers this. And in general, it rather points to you having invalid data or mapping, that function should work just fine in both drivers fine - it is there to adjust how dates are logged in mongo (but it works on prototype level, so if you use both drivers, it will affect the sql driver too - again, just the logging).

It's not related, but you **should not set `allowGlobalContext: true`**, this will backfire, things won't work in parallel properly. [Don't disable validations you don't understand!](https://stackoverflow.com/a/72799993/3665878)

**Reply by Ahmadjlallad** [reply]
Thank you! I'd love to share the repository with you, but it's a private repository on Bitbucket for the company :[ . As for the allowGlobalContext: true setting, yes, I was experimenting with it to see what would happen, but I've figured it out now, thank you. The issue is exactly as you described; I was logging both the query result and its return. This is an older database, and some MySQL mode validations are disabled, such as [no zero in date](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_zero_in_date). Perhaps this could affect the result.

what do you think?

**Reply by B4nan** [reply]
It it happens locally, try to put some logging into the place that fails to see how the value looks like.
